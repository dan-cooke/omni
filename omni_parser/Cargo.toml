[package]
name = "omni_parser"
version = "0.1.0"
authors = ["Daniel Cooke <danielcooke1996@gmail.com"]
edition = "2021"

[build-dependencies]
lalrpop = "0.20.0"

[dependencies]
lalrpop-util = { version = "0.20.0", features = ["lexer"] }
logos = "0.13.0"
serde = { version = "1.0.193", features = ["derive"] }
serde_json = "1.0.108"

[dev-dependencies]
insta = { version = "1.34.0", features = ["json"] }




