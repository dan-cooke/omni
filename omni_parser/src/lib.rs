pub mod ast;
mod ast_util;
mod errors;
mod helpers;
mod lexer;
mod parser;
mod tokens;

pub use parser::*;
