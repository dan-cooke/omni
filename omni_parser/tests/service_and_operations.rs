use insta::assert_json_snapshot;
use omni_parser::*;

#[test]
pub fn services_can_be_parsed() {
    let ast = parse_file("tests/test-schemas/types/service_and_operations.omni");

    assert_json_snapshot!(ast);
}
