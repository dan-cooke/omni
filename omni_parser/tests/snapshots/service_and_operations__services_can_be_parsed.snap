---
source: omni_parser/tests/service_and_operations.rs
expression: ast
---
{
  "body": [
    {
      "ServiceDef": {
        "id": {
          "name": "TestService",
          "span": {
            "file_name": "tests/test-schemas/types/service_and_operations.omni",
            "start": 8,
            "end": 19
          }
        },
        "properties": [
          {
            "id": {
              "name": "version",
              "span": {
                "file_name": "tests/test-schemas/types/service_and_operations.omni",
                "start": 24,
                "end": 31
              }
            },
            "value": {
              "Literal": {
                "value": {
                  "String": "\"0.1.0\""
                },
                "span": {
                  "file_name": "tests/test-schemas/types/service_and_operations.omni",
                  "start": 33,
                  "end": 40
                }
              }
            },
            "span": {
              "file_name": "tests/test-schemas/types/service_and_operations.omni",
              "start": 24,
              "end": 40
            }
          },
          {
            "id": {
              "name": "operations",
              "span": {
                "file_name": "tests/test-schemas/types/service_and_operations.omni",
                "start": 44,
                "end": 54
              }
            },
            "value": {
              "ArrayExpression": {
                "elements": [
                  {
                    "name": "TestOperation",
                    "span": {
                      "file_name": "tests/test-schemas/types/service_and_operations.omni",
                      "start": 57,
                      "end": 70
                    }
                  }
                ],
                "span": {
                  "file_name": "tests/test-schemas/types/service_and_operations.omni",
                  "start": 56,
                  "end": 71
                }
              }
            },
            "span": {
              "file_name": "tests/test-schemas/types/service_and_operations.omni",
              "start": 44,
              "end": 71
            }
          }
        ],
        "span": {
          "file_name": "tests/test-schemas/types/service_and_operations.omni",
          "start": 0,
          "end": 73
        }
      }
    },
    {
      "StructDef": {
        "id": {
          "name": "TestOperationInput",
          "span": {
            "file_name": "tests/test-schemas/types/service_and_operations.omni",
            "start": 82,
            "end": 100
          }
        },
        "properties": [
          {
            "id": {
              "name": "start",
              "span": {
                "file_name": "tests/test-schemas/types/service_and_operations.omni",
                "start": 105,
                "end": 110
              }
            },
            "value": {
              "Identifier": {
                "name": "Timestamp",
                "span": {
                  "file_name": "tests/test-schemas/types/service_and_operations.omni",
                  "start": 112,
                  "end": 121
                }
              }
            },
            "span": {
              "file_name": "tests/test-schemas/types/service_and_operations.omni",
              "start": 105,
              "end": 121
            }
          },
          {
            "id": {
              "name": "end",
              "span": {
                "file_name": "tests/test-schemas/types/service_and_operations.omni",
                "start": 125,
                "end": 128
              }
            },
            "value": {
              "Identifier": {
                "name": "Timestamp",
                "span": {
                  "file_name": "tests/test-schemas/types/service_and_operations.omni",
                  "start": 130,
                  "end": 139
                }
              }
            },
            "span": {
              "file_name": "tests/test-schemas/types/service_and_operations.omni",
              "start": 125,
              "end": 139
            }
          }
        ],
        "span": {
          "file_name": "tests/test-schemas/types/service_and_operations.omni",
          "start": 75,
          "end": 141
        }
      }
    },
    {
      "StructDef": {
        "id": {
          "name": "TestOperationOutput",
          "span": {
            "file_name": "tests/test-schemas/types/service_and_operations.omni",
            "start": 150,
            "end": 169
          }
        },
        "properties": [
          {
            "id": {
              "name": "data",
              "span": {
                "file_name": "tests/test-schemas/types/service_and_operations.omni",
                "start": 174,
                "end": 178
              }
            },
            "value": {
              "Identifier": {
                "name": "String",
                "span": {
                  "file_name": "tests/test-schemas/types/service_and_operations.omni",
                  "start": 180,
                  "end": 186
                }
              }
            },
            "span": {
              "file_name": "tests/test-schemas/types/service_and_operations.omni",
              "start": 174,
              "end": 186
            }
          }
        ],
        "span": {
          "file_name": "tests/test-schemas/types/service_and_operations.omni",
          "start": 143,
          "end": 189
        }
      }
    },
    {
      "OperationDef": {
        "id": {
          "name": "TestOperation",
          "span": {
            "file_name": "tests/test-schemas/types/service_and_operations.omni",
            "start": 201,
            "end": 214
          }
        },
        "properties": [
          {
            "id": {
              "name": "input",
              "span": {
                "file_name": "tests/test-schemas/types/service_and_operations.omni",
                "start": 219,
                "end": 224
              }
            },
            "value": {
              "Identifier": {
                "name": "TestOperationInput",
                "span": {
                  "file_name": "tests/test-schemas/types/service_and_operations.omni",
                  "start": 226,
                  "end": 244
                }
              }
            },
            "span": {
              "file_name": "tests/test-schemas/types/service_and_operations.omni",
              "start": 219,
              "end": 244
            }
          },
          {
            "id": {
              "name": "output",
              "span": {
                "file_name": "tests/test-schemas/types/service_and_operations.omni",
                "start": 248,
                "end": 254
              }
            },
            "value": {
              "Identifier": {
                "name": "TestOperationOutput",
                "span": {
                  "file_name": "tests/test-schemas/types/service_and_operations.omni",
                  "start": 256,
                  "end": 275
                }
              }
            },
            "span": {
              "file_name": "tests/test-schemas/types/service_and_operations.omni",
              "start": 248,
              "end": 275
            }
          }
        ],
        "span": {
          "file_name": "tests/test-schemas/types/service_and_operations.omni",
          "start": 191,
          "end": 277
        }
      }
    }
  ]
}
